File: Day47/NoOfSubStringsContainingAll3Characters.java
import java.util.*;

public class NoOfSubStringsContainingAll3Characters {
    public int numberOfSubstrings(String s) {
        int count = 0;
        int lastA = -1, lastB = -1, lastC = -1;
        for (int i = 0; i < s.length(); i++) {
            char c = s.charAt(i);
            switch (c) {
                case 'a': lastA = i; break;
                case 'b': lastB = i; break;
                case 'c': lastC = i; break;
            }
            if (lastA != -1 && lastB != -1 && lastC != -1) {
                int minLast = Math.min(lastA, Math.min(lastB, lastC));
                count += minLast + 1;
            }
        }
        return count;
    }

    public static void main(String[] args) {
        Scanner sc = new Scanner(System.in);
        String s = sc.nextLine();
        NoOfSubStringsContainingAll3Characters sol = new NoOfSubStringsContainingAll3Characters();
        System.out.println(sol.numberOfSubstrings(s));
    }
}

File: Day47/NoOfSubStringsContainingAll3Characters.py
class Solution(object):
    def numberOfSubstrings(self, s):
        count = 0
        last_a = last_b = last_c = -1
        for i, char in enumerate(s):
            if char == 'a':
                last_a = i
            elif char == 'b':
                last_b = i
            else:
                last_c = i
            if last_a != -1 and last_b != -1 and last_c != -1:
                min_last = min(last_a, last_b, last_c)
                count += min_last + 1
        return count

if __name__ == "__main__":
    s = input().strip()
    sol = Solution()
    print(sol.numberOfSubstrings(s))
